### PR TITLE
[To Do] Optimize show To Do flow based on Quirine's latest design

### LINKED ISSUE #741: [TodoSkill] To Do pagination shows first page twice
**Project**
Skills, TodoSkill

**Description**
The first page shows twice when paginating over the a list.

**To Reproduce**
1. Create a list with more items than the ones that can be displayed on a page. In this version the page size is items so I created a To Do list with 8 items
2. Say show my to do (and the first 6 items will show up)
3. Say yes when prompted to see more. The same page is repeated

**Expected behavior**
I would expect pagination to work.

**Screenshots**

image

**Additional context**
Also the text in the chat window says "Do you want to **hear** more from your list?" In the text window it should say "Do you want to **see** more from your list?" or "Do you want to **read** more from your list?" not "hear", hear is fine for the Speak response but not for text.

### LINKED ISSUE #787: [To Do]Adjust responses to differentiate the wording of speech and text
**Project**
Skills

**Description**
Adjust responses to differentiate the wording of speech and text

**To Reproduce**
Current some responses have same speech and text string which is confusing users.

**Expected behavior**
The responses have different strings in speech and text mode.

**Screenshots**

**Additional context**

### LINKED ISSUE #805: [TodoSkill] Mark X complete doesn't detect the entity
**Project**
Skills

**Description**
When I say "mark *xTask* completed* the bot prompts me for the task name, it doesn't detect the entity.

**To Reproduce**
1. Add meat to your grocery list
2. The bot will prompt you for the task name, it should have inferred that from the utterance.
image
image

**Expected behavior**
I would expect the bot to mark meat as completed and be done with it.

### PR DESCRIPTION
**Pull request by feich-ms**
Optimize show To Do flow based on Quirine's latest design
## Description
1. Redesign the read size and display size logic. Remove read size from config and keep read size same as display size.
2, Optimize some responses based on the new design.
3. Optimize some responses to distiguish the differences of speech and text.
4. Optimize some corner cases flow like "show more after last page" and "show previous at first page" based on the new design.
5. Adjust the unit tests to align with the new flow

## Related Issue
fixes #741 
fixes #787 
fixes #805 

## Testing Steps
1. Ask "show to do list" and then say "show next" to see and hear the next page tasks.
2. Ask "show next" to next page
3. Ask "show previous" to first page

## Checklist
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have added or updated the appropriate unit tests
- [x] I have tested any new/updated dialogs using Speech in the emulator to ensure the [Speak](https://docs.microsoft.com/en-us/azure/bot-service/dotnet/bot-builder-dotnet-text-to-speech?view=azure-bot-service-3.0) property is set to enable a high quality speech-first experience and the appropriate [InputHints](https://docs.microsoft.com/en-us/azure/bot-service/dotnet/bot-builder-dotnet-add-input-hints?view=azure-bot-service-3.0) are set correctly.

### CODE REVIEW

File: solutions/Virtual-Assistant/src/csharp/skills/todoskill/todoskill/Dialogs/Shared/Resources/ToDoSharedResponses.json [hunk 1]
```diff
@@ -116,6 +116,15 @@
     ],
     "inputHint": "ignoringInput"
   },
+  "CardSummaryMessageForSingleTask": {
```
Comment (lzc850612) [inline solutions/Virtual-Assistant/src/csharp/skills/todoskill/todoskill/Dialogs/Shared/Resources/ToDoSharedResponses.json:119]: is this response only gonna be used for single task? if yes then it should just be 'You have 1 task on your {listType} list'. and you need to have a response for multiple task as well.

File: solutions/Virtual-Assistant/src/csharp/skills/todoskill/todoskill/Dialogs/ShowToDo/Resources/ShowToDoResponses.cs [hunk 2]
```diff
@@ -12,23 +12,27 @@ namespace ToDoSkill.Dialogs.ShowToDo.Resources
     public class ShowToDoResponses : IResponseIdCollection
     {
         // Generated accessors
-		public const string LatestOneTask = "LatestOneTask";
-		public const string LatestTwoTasks = "LatestTwoTasks";
-		public const string LatestThreeOrMoreTasks = "LatestThreeOrMoreTasks";
+		public const string LatestTask = "LatestTask";
```
Comment (lzc850612) [inline solutions/Virtual-Assistant/src/csharp/skills/todoskill/todoskill/Dialogs/ShowToDo/Resources/ShowToDoResponses.cs:15]: weird indentation?

File: solutions/Virtual-Assistant/src/csharp/skills/todoskill/todoskill/Dialogs/ShowToDo/Resources/ShowToDoResponses.json [hunk 3]
```diff
@@ -38,116 +38,152 @@
...
-        "text": "The next tasks are",
-        "speak": "The next tasks are"
+        "text": "The next tasks are ",
+        "speak": "The next tasks are "
       }
     ],
     "inputHint": "ignoringInput"
   },
-  "NextThreeOrMoreTask": {
+  "PreviousTasks": {
```
Comment (lzc850612) [inline solutions/Virtual-Assistant/src/csharp/skills/todoskill/todoskill/Dialogs/ShowToDo/Resources/ShowToDoResponses.json:92]: wondering if it's necessary to have separate responses for single and plural cases

File: solutions/Virtual-Assistant/src/csharp/skills/todoskill/todoskill/Dialogs/ShowToDo/ShowToDoItemDialog.cs [hunk 4]
```diff
@@ -345,8 +344,8 @@ public async Task<DialogTurnResult> AskSecondReadMoreConfirmation(WaterfallStepC
         {
             try
             {
-                var prompt = ResponseManager.GetResponse(ShowToDoResponses.ReadMoreTasksPrompt);
-                var retryPrompt = ResponseManager.GetResponse(ShowToDoResponses.ReadMoreTasksConfirmFailed);
+                var prompt = ResponseManager.GetResponse(ShowToDoResponses.ReadMoreTasksPrompt2);
```
Comment (lzc850612) [inline solutions/Virtual-Assistant/src/csharp/skills/todoskill/todoskill/Dialogs/ShowToDo/ShowToDoItemDialog.cs:347]: should put different ways of saying the responses in the same response config in the response file, and have the responsemanager handle the randomization of responses.

**Review by lzc850612 — CHANGES_REQUESTED** [review]

File: solutions/Virtual-Assistant/src/csharp/skills/todoskill/todoskill/Dialogs/ShowToDo/Resources/ShowToDoResponses.cs (hunk 2 above)
Comment (feich-ms) [reply solutions/Virtual-Assistant/src/csharp/skills/todoskill/todoskill/Dialogs/ShowToDo/Resources/ShowToDoResponses.cs:15]: @lzc850612 Hi Ted, what do you mean "weird indentation" here? "LastestTask" means latest single task and "LatestTasks" means latest multiple tasks. "LatestThreeOrMoreTasks" is no longer used in new flow.

File: solutions/Virtual-Assistant/src/csharp/skills/todoskill/todoskill/Dialogs/ShowToDo/Resources/ShowToDoResponses.json (hunk 3 above)
Comment (feich-ms) [reply solutions/Virtual-Assistant/src/csharp/skills/todoskill/todoskill/Dialogs/ShowToDo/Resources/ShowToDoResponses.json:92]: @lzc850612 Hi Ted, here "PreviousTasks" will not have "single" case as the response "Your previous tasks are" is only used in ShowPrevious flow in which tasks count in previous page must be display size(current is 4).

File: solutions/Virtual-Assistant/src/csharp/skills/todoskill/todoskill/Dialogs/ShowToDo/ShowToDoItemDialog.cs (hunk 4 above)
Comment (feich-ms) [reply solutions/Virtual-Assistant/src/csharp/skills/todoskill/todoskill/Dialogs/ShowToDo/ShowToDoItemDialog.cs:347]: @lzc850612 From Quirine's design, these two reponses(ReadMoreTasksPrompt and ReadMoreTasksPrompt2) are used in different scenarios. The first one is only used in the first ReadMore prompt and the second is used in the following ReadMore prompts, so I differentiate them.

File: solutions/Virtual-Assistant/src/csharp/skills/todoskill/todoskill/Dialogs/Shared/Resources/ToDoSharedResponses.json (hunk 1 above)
Comment (feich-ms) [reply solutions/Virtual-Assistant/src/csharp/skills/todoskill/todoskill/Dialogs/Shared/Resources/ToDoSharedResponses.json:119]: @lzc850612 Yes, it is only used for single task. Changed per your suggestion. I have the response called "CardSummaryMessage" for multiple tasks. The name is not friendly and I changed it to "CardSummaryMessageForMultipleTasks".

File: solutions/Virtual-Assistant/src/csharp/skills/todoskill/todoskill/Dialogs/ShowToDo/Resources/ShowToDoResponses.cs (hunk 2 above)
Comment (lzc850612) [reply solutions/Virtual-Assistant/src/csharp/skills/todoskill/todoskill/Dialogs/ShowToDo/Resources/ShowToDoResponses.cs:15]: yeah i just mean that the code here has two tabs on the left, not aligned with the comment

File: solutions/Virtual-Assistant/src/csharp/skills/todoskill/todoskill/Dialogs/ShowToDo/ShowToDoItemDialog.cs (hunk 4 above)
Comment (lzc850612) [reply solutions/Virtual-Assistant/src/csharp/skills/todoskill/todoskill/Dialogs/ShowToDo/ShowToDoItemDialog.cs:347]: ok

**Review by lzc850612 — APPROVED** [review]

File: solutions/Virtual-Assistant/src/csharp/skills/todoskill/todoskill/Dialogs/ShowToDo/Resources/ShowToDoResponses.cs (hunk 2 above)
Comment (feich-ms) [reply solutions/Virtual-Assistant/src/csharp/skills/todoskill/todoskill/Dialogs/ShowToDo/Resources/ShowToDoResponses.cs:15]: The code is automatically generated by template ResponseIdCollection.t4.
